Add support to write multidimensional string arrays

stephprince opened the pull request:
## Motivation

Fixes https://github.com/hdmf-dev/hdmf/issues/1137 and fixes https://github.com/hdmf-dev/hdmf/issues/1096. 

On build, multidimensional string lists/arrays were being converted into 1-D lists/arrays:

`[['aa', 'bb'], ['cc', 'dd']] -> ["['aa', 'bb']", "['cc', 'dd']"]`.

On write, the shape of a multidimensional string dataset was being defined incorrectly and so the dataset could not be written. The data shape definition was being caught by this conditional that I believe is intended to check for structured arrays / compound data types:

https://github.com/hdmf-dev/hdmf/blob/7426275cacc769a10ffca89836765df1355ba9db/src/hdmf/backends/hdf5/h5tools.py#L1472-L1473

 I updated the conditional and `get_data_shape` is now used to set the shape of the dataset for multidimensional string lists/arrays.



## How to test the behavior?

This came up in https://github.com/NeurodataWithoutBorders/pynwb/pull/1924. 

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [X] Does the PR clearly describe the problem and the solution?
- [X] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [X] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

stephprince committed:
add condition for multidim string arrays

stephprince committed:
add tests for multidim string array build

stephprince committed:
update condition when defining hdf5 dataset shape

stephprince committed:
add test to write multidim string array

stephprince committed:
update CHANGELOG.md

stephprince committed:
fix text decoding in test

rly commented on src/hdmf/build/objectmapper.py:
```diff
@@ -617,7 +617,9 @@ def __convert_string(self, value, spec):
                         return x.isoformat()  # method works for both date and datetime
                 if string_type is not None:
-                    if spec.shape is not None or spec.dims is not None:
+                    if spec.shape is not None and len(spec.shape) > 1:
+                        ret = [[string_type(item) for item in sublist] for sublist in value]
```
This is fine for 2D arrays but I think it will fail for higher-dimensional arrays. Can you add a test for a 3D array? You could convert value to a numpy array and change the type. Or do the conversion recursively on the list elements.

stephprince replied:
Good point. Converting it to a numpy array and changing the type didn't work as well with the isodatetimes so I added a recursive conversion.

stephprince committed:
add recursive string type for arrays of arbitrary dim

stephprince committed:
add test for compound data type with strings

stephprince committed:
add tests for multidim str attributes

stephprince committed:
fix line lengths

stephprince commented:
@rly I realized that I had only updated the string conversion for datasets so I updated `__convert_string` to accommodate multidimensional string arrays as attributes as well. 

For testing, I added an additional `attr_array` argument to the `Bar` class constructor, but I'm not sure if that was the best approach or if there's a better way to test it since the `Bar` class seemed to be used in multiple places.

stephprince commented on tests/unit/test_io_hdf5_h5tools.py:
```diff
@@ -164,6 +164,30 @@ def test_write_dataset_list(self):
+                        for sublist in dset[:]]
+        self.assertTrue(decoded_dset == a)
+
+    def test_write_dataset_list_compound_datatype(self):
```
There was missing code coverage for compound data types in the `__list_fill__` method in `src/hdmf/backends/hdf5/h5tools.py` so I added this test that I think should cover that case.

rly replied:
Looks good!

stephprince committed:
Merge branch 'dev' into write-multidim-str-arrays

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stephprince committed:
update compound dtype test

rly committed:
Merge branch 'dev' into write-multidim-str-arrays
